src/validation.cpp

```diff
@@ -5045,12 +5046,30 @@ bool LoadMempool(CTxMemPool& pool)
         for (const auto& i : mapDeltas) {
             pool.PrioritiseTransaction(i.first, i.second);
         }
+
+        // check if there are unbroadcast transactions
+        bool has_unbroadcast_txns = false;
+
+        try {
+            file >> has_unbroadcast_txns;
+        } catch (const std::exception &e) {
+            // eof gets set when we try to read the end of the file
+            // so specifically catch expected error and move on.
+            if (!file.eof()) throw e;
+        }
+
+        if (has_unbroadcast_txns) {
```

Comment:
in commit b0f6a238335d96c9f5f7c2b83aabd56ab4f47a71:

Why is this needed?

I think you can just go ahead and assume that the set was serialized. If it wasn't it will result in a one-time `LogPrintf` that the mempool data couldn't be deserialized. I think this is fine if it is mentioned in the release notes or as a code comment or not at all.

Reply:
There's no LogPrintf there though?

Reply:
so, are you proposing 
- remove writing the boolean `has_unbroadcast_txns` to the file [here](https://github.com/bitcoin/bitcoin/pull/18038/files#diff-24efdb00bfbe56b140fb006b562cc70bR5121)
- remove try block above, just leave the `file>>pool.m_unbroadcast_txids;` code below 
- if there are no `m_unbroadcast_txids`, there will be a `LogPrintf` that "Failed to deserialize mempool data on disk. Continuing anyway", and code moves on? 

if I am interpreting correctly.... I find that to be very misleading user output. And I'd expect it to happen frequently since the unbroadcast set is more of a safety mechanism and hopefully nodes aren't often shutting down without their local transactions being broadcast to the network. So I'm not really seeing the reasoning / benefits. But maybe I'm misunderstanding your suggestion.

Reply:
The logprint should only happen on update. A serialized empty set is hopefully different from no serialized set.

Reply:
See here:

https://github.com/bitcoin/bitcoin/blob/5236b2e267a58870239673c7ec85e5df0cb8fc8e/src/serialize.h#L972

Reply:
Yeah, so in normal case we'll get this error once: when the node is updated to use this code, and attempts to use the file from previous runs. Which is, yeah, not ideal, but probably fine.
If you'd really want to avoid that, I'd rather add versioning of the file, rather than having bools like that you suggest.
(See how we do versioning in peers.dat AddrMan serialization if interested)

Reply:
This file is already versioned, but the version shouldn't be bumped because that will silently drop *all* transactions of the mempool

Reply:
Why does it have to drop all transactions? The new code can just be aware of the previous version, and not attempt to read unbroadcast_txs in that case, but just carefully read all old transactions (the advantage over this current PR is that no error is thrown/logged).

Reply:
> Why does it have to drop all transactions?

When bumping the version here and assuming the code is merged into 0.21.0, then reading a mempool.dat created by Bitcoin Core 0.21.0 will fail completely on a Bitcoin Core 0.20.0, because the code is already released and the logic of that code is to drop all transactions when the version is not equal to 1.

Reply:
True. But who would use a mempool.dat from 0.21 on 0.20 node? I would say the most common scenario is opposite, upgrading from 0.20 to 0.21.

Reply:
Miners might want to upgrade, test the new version for a bit and then downgrade temporarily because they noticed an RPC breaking change. They should be able to continue mining on the txs they had before. For many users, a version upgrade/downgrade is the only time they restart the node. So this is the only time when this code is run and can provide them with value. If the code fails to be useful in the only use-case we might as well remove it altogether. ;)

Reply:
If you guys think that a try-catch is useful to avoid the debug LogPrint, fine. I am ok with that too. But I don't see a reason for this boolean.

Reply:
I also don't see a good reason for this boolean.

Reply:
ok! I'm following now. I wasn't aware of the serialization logic. I'll remove the bool & try-catch as well. thanks!

Reply:
fixed! thanks again

Reply:
> but the version shouldn't be bumped because that will silently drop all transactions of the mempool

@MarcoFalke - this is interesting. It means if we ever have to bump the version number we'd have to do something like:

- Bitcoin Core version X only understands mempool.dat v1
- add mempool.dat v2 parsing logic in Bitcoin Core version X+1
- add mempoo.dat v2 writing logic in Bitcoin Core version X+2

I think to avoid the debug LogPrint we can wrap the m_unbroadcast_txids parsing in its own try-except. Something like:

```
        try {
            LOCK(pool.cs);
            file >> pool.m_unbroadcast_txids;
            unbroadcast = pool.m_unbroadcast_txids.size();
        } catch (const std::exception& e) {
            // mempool.dat files created prior to v0.21 will not have an
            // unbroadcast set. No need to log a failure if parsing fails
            // here.
        }
```

Reply:
> I think to avoid the debug LogPrint we can wrap the m_unbroadcast_txids parsing in its own try-except

I considered doing this but decided not to because a one-time error message (with functionality working fine) seems more desirable than a nested try / catch statement that would probably live in the code for a very long time.

Reply:
ah, based on another review comment or two, ended up implementing here 191fc04e9bb99589f204ccfceece7d715a82f4c0 (PR #18807)